protocol/daemons/liquidation/client/client.go

```diff
@@ -65,262 +80,56 @@ func Start(
+	client *Client,
+	s SubTaskRunner,
+	flags flags.DaemonFlags,
+	ticker *time.Ticker,
+	done <-chan bool,
+	subaccountQueryClient satypes.QueryClient,
+	clobQueryClient clobtypes.QueryClient,
+	liquidationServiceClient api.LiquidationServiceClient,
+	logger log.Logger,
 ) {
-	defer telemetry.ModuleMeasureSince(metrics.LiquidationDaemon, time.Now(), metrics.GetAllSubaccounts, metrics.Latency)
-	subaccounts = make([]satypes.Subaccount, 0)
-
-	var nextKey []byte
 	for {
-		subaccountsFromKey, next, err := getSubaccountsFromKey(
-			ctx,
-			client,
-			limit,
-			nextKey,
-		)
-
-		if err != nil {
-			return nil, err
-		}
-
-		subaccounts = append(subaccounts, subaccountsFromKey...)
-		nextKey = next
-
-		if len(nextKey) == 0 {
-			break
-		}
-	}
-
-	telemetry.ModuleSetGauge(
-		metrics.LiquidationDaemon,
-		float32(len(subaccounts)),
-		metrics.GetAllSubaccounts,
-		metrics.Count,
-	)
-
-	return subaccounts, nil
-}
-
-// GetLiquidatableSubaccountIds verifies collateralization statuses of subaccounts with
-// at least one open position and returns a list of unique and potentially liquidatable subaccount ids.
-func GetLiquidatableSubaccountIds(
-	ctx context.Context,
-	client clobtypes.QueryClient,
-	liqFlags flags.LiquidationFlags,
-	subaccounts []satypes.Subaccount,
-) (
-	liquidatableSubaccountIds []satypes.SubaccountId,
-	err error,
-) {
-	defer telemetry.ModuleMeasureSince(
-		metrics.LiquidationDaemon,
-		time.Now(),
-		metrics.GetLiquidatableSubaccountIds,
-		metrics.Latency,
-	)
-
-	// Filter out subaccounts with no open positions.
-	subaccountsToCheck := make([]satypes.SubaccountId, 0)
-	for _, subaccount := range subaccounts {
-		if len(subaccount.PerpetualPositions) > 0 {
-			subaccountsToCheck = append(subaccountsToCheck, *subaccount.Id)
-		}
-	}
-
-	telemetry.ModuleSetGauge(
-		metrics.LiquidationDaemon,
-		float32(len(subaccountsToCheck)),
-		metrics.SubaccountsWithOpenPositions,
-		metrics.Count,
-	)
-
-	// Query the gRPC server in chunks of size `liqFlags.RequestChunkSize`.
-	liquidatableSubaccountIds = make([]satypes.SubaccountId, 0)
-	for start := 0; start < len(subaccountsToCheck); start += int(liqFlags.RequestChunkSize) {
-		end := lib.Min(start+int(liqFlags.RequestChunkSize), len(subaccountsToCheck))
-
-		results, err := CheckCollateralizationForSubaccounts(
-			ctx,
-			client,
-			subaccountsToCheck[start:end],
-		)
-		if err != nil {
-			return nil, err
-		}
-
-		for _, result := range results {
-			if result.IsLiquidatable {
-				liquidatableSubaccountIds = append(liquidatableSubaccountIds, result.SubaccountId)
+		select {
+		case <-ticker.C:
+			if err := s.RunLiquidationDaemonTaskLoop(
+				ctx,
+				flags.Liquidation,
+				subaccountQueryClient,
+				clobQueryClient,
+				liquidationServiceClient,
+			); err != nil {
+				// TODO(DEC-947): Move daemon shutdown to application.
+				logger.Error("Liquidations daemon returned error", "error", err)
+				client.ReportFailure(err)
+			} else {
+				client.ReportSuccess()
```

Comment:
for my understanding - is the client still reporting to the server? how is this implementation fundamentally different from the previous implementation?

If a task loop takes more than 5 minutes, we still report an error, right? this feels like we are effectively just setting the timeout from 3 * 1.6s = 4.8s to 5 minutes.

Reply:
Yes, if the task loop takes more than 5 minutes to complete, the daemon will enter into an unhealthy state until the update finishes - so if it extends to 6 minutes, then the daemon would be unhealthy for 1 minute. I added this implementation detail in order to capture cases where a daemon could hang/deadlock after a successful update.

I can take it out, but would ask:
- do we realistically expect the daemon to ever run for 5 minutes and, if so, would that have any deleterious effect on the protocol?
- is there a duration threshold that could work better here? It was meant to be a large enough number to be unrealistic but to capture the hanging case so that we don't silently fail, which would be a false positive for daemon health. 10 minutes or even half an hour is fine - I just want to capture this case, while also hopefully informing validators earlier than later if their nodes aren't able to make accurate proposals.

Reply:
Note too - if the daemon does become unhealthy, we are still adding a flag that allows users to extend the period of unhealth before the protocol panics from 5 minutes to indefinite.

Reply:
As discussed offline, this PR removes the staleness check in the health-checkable and also reports success whenever paginated gRPC queries return successfully.